File: api/views/acl/permission.py
# -*- coding:utf-8 -*-


from flask import request

from api.lib.decorator import args_required
from api.lib.perm.acl.permission import PermissionCRUD
from api.lib.utils import handle_arg_list
from api.resource import APIView


class ResourcePermissionView(APIView):
    url_prefix = ("/resources/<int:resource_id>/permissions", "/resource_groups/<int:group_id>/permissions")

    def get(self, resource_id=None, group_id=None):
        return self.jsonify(PermissionCRUD.get_all(resource_id, group_id))


class RolePermissionGrantView(APIView):
    url_prefix = ('/roles/<int:rid>/resources/<int:resource_id>/grant',
                  '/roles/<int:rid>/resource_groups/<int:group_id>/grant')

    @args_required('perms')
    def post(self, rid, resource_id=None, group_id=None):
        perms = handle_arg_list(request.values.get("perms"))
        PermissionCRUD.grant(rid, perms, resource_id=resource_id, group_id=group_id)

        return self.jsonify(rid=rid, resource_id=resource_id, group_id=group_id, perms=perms)


class RolePermissionRevokeView(APIView):
    url_prefix = ('/roles/<int:rid>/resources/<int:resource_id>/revoke',
                  '/roles/<int:rid>/resource_groups/<int:group_id>/revoke')

    @args_required('perms')
    def post(self, rid, resource_id=None, group_id=None):
        perms = handle_arg_list(request.values.get("perms"))
        PermissionCRUD.revoke(rid, perms, resource_id=resource_id, group_id=group_id)

        return self.jsonify(rid=rid, resource_id=resource_id, group_id=group_id, perms=perms)

File: cmdb-ui/src/api/acl/permission.js
import { axios } from '@/utils/request'

const urlPrefix = '/v1/acl'

export function getResourcePerms (resourceID) {
  return axios({
    url: urlPrefix + `/resources/${resourceID}/permissions`,
    method: 'GET'
  })
}

export function getResourceTypePerms (typeID) {
  return axios({
    url: urlPrefix + `/resource_types/${typeID}/perms`,
    method: 'GET'
  })
}

export function getResourceGroupPerms (resourceGroupID) {
  return axios({
    url: urlPrefix + `/resource_groups/${resourceGroupID}/permissions`,
    method: 'GET'
  })
}

export function setRoleResourcePerm (rid, resourceID, params) {
  return axios({
    url: urlPrefix + `/roles/${rid}/resources/${resourceID}/grant`,
    method: 'POST',
    data: params
  })
}

export function setRoleResourceGroupPerm (rid, resourceGroupID, params) {
  return axios({
    url: urlPrefix + `/roles/${rid}/resource_groups/${resourceGroupID}/grant`,
    method: 'POST',
    data: params
  })
}

export function deleteRoleResourcePerm (rid, resourceID, params) {
  return axios({
    url: urlPrefix + `/roles/${rid}/resources/${resourceID}/revoke`,
    method: 'POST',
    data: params
  })
}

export function deleteRoleResourceGroupPerm (rid, resourceGroupID, params) {
  return axios({
    url: urlPrefix + `/roles/${rid}/resource_groups/${resourceGroupID}/revoke`,
    method: 'POST',
    data: params
  })
}
